src/skirmishbunker/FlatRoof.py

```diff
@@ -1,59 +1,338 @@
         self.roof_chamfer_edges_selector = ""
-        self.roof_chamfer = 10
+        self.roof_chamfer = 0
         self.roof_operation = "chamfer" # chamfer, fillet
 
+        # roof tiles
+        self.render_tiles = False
+        self.tile_size = 21
+        self.tile_padding = 2
+        self.tile_height = 1.5
+        # less then -1 results in a cut
+        self.tile_z_offset = -1
+
+        # Hatches
+        self.hatch_panels = []
+        self.hatch_length = 25
+        self.hatch_width = 25
+        self.hatch_radius = 10.5
+        self.hatch_height = 6
+        self.hatch_z_translate = 0
+
+        # Pip/Magnet holes
+        self.cut_holes = False
+        self.hole_inset = 1.5
+        self.hole_depth = 1
+        self.hole_diameter = 2
+
         #shapes
         self.roof_body = None
+        self.tiles = None
+        self.hatches = None
+        self.holes = None
 
-    def __make_roof_body(self):
-        roof_body = (
-            cq.Workplane("XY")
-            .box(self.length,self.width,self.height)
-        )
+    def should_cut_tiles(self):
+        if self.tile_z_offset < -1:
+            return True
+
+        return False
 
-        if self.roof_chamfer and self.roof_chamfer < self.height:
-            if self.roof_operation.lower() == "chamfer":
-                roof_body = (
-                    roof_body
-                    .faces(self.roof_chamfer_faces_selector)
-                    .edges(self.roof_chamfer_edges_selector)
-                    .chamfer(self.roof_chamfer)
-                )
-            elif self.roof_operation.lower() == "fillet":
-                roof_body = (
-                    roof_body
-                    .faces(self.roof_chamfer_faces_selector)
-                    .edges(self.roof_chamfer_edges_selector)
-                    .fillet(self.roof_chamfer)
-                )
-            else:
-                raise Exception("Unrecognied roof operation")
+    def calc_final_length(self):
+        return self.length - (self.inset * 2)
+
+    def calc_final_width(self):
+        return self.width - (self.inset * 2)
+
+    def calc_tile_space_length(self):
+        length = self.length
+        length -= 2 * self.inset
+        length -= 2 * self.roof_chamfer
+        return length
+
+    def calc_tile_space_width(self):
+        width = self.width
+        width -= 2 * self.inset
+        width -= 2 * self.roof_chamfer
+        return width
+
+    def calc_tile_spacing(self):
+        return self.tile_size + self.tile_padding
+
+    def calc_tile_z_translate(self):
+        cut_tiles = self.should_cut_tiles()
+        translate = (self.height / 2 + self.tile_height / 2)
+
+        if cut_tiles == True:
+            translate = self.height - translate
         else:
-            # better to kill if this condition is met otherwise it fails silently.
+            translate = translate
+        
+        return translate
+
+    def calc_hatch_space_length(self):
```

Comment:
If these are private then they cannot be seen by inheritors. If you check what I have done with the detail roof, it inherits heavily from it's parent.

They could be protected instead of public though... 🤷.